File: src/hilbert/models/finding.py
"""Finding models for Hilbert."""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


class Finding(BaseModel):
    """A finding extracted from a paper during research."""

    finding_id: str
    claim: str = Field(description="The claim or finding statement")
    source_paper_id: str = Field(description="ID of the source paper")
    evidence_text: str = Field(description="Text from paper supporting the claim")
    confidence: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
        description="Confidence score from verifier (0.0-1.0)"
    )
    created_at: datetime = Field(default_factory=datetime.now)
    is_verified: bool = False

    def confidence_label(self) -> str:
        """Human-readable confidence label."""
        if self.confidence >= 0.9:
            return "high"
        elif self.confidence >= 0.75:
            return "medium"
        elif self.confidence > 0:
            return "low"
        return "unverified"


class Gap(BaseModel):
    """A gap in coverage identified by the reviewer."""

    gap_id: str
    description: str
    sub_question: Optional[str] = None
    severity: str = "minor"  # minor, major, critical
    related_findings: list[str] = Field(default_factory=list)

File: cli/src/core/models/finding.ts
import { z } from "zod";

export const FindingSchema = z.object({
  finding_id: z.string(),
  claim: z.string().describe("The claim or finding statement"),
  source_paper_id: z.string().describe("ID of the source paper"),
  evidence_text: z.string().describe("Text from paper supporting the claim"),
  confidence: z
    .number()
    .min(0)
    .max(1)
    .default(0)
    .describe("Confidence score from verifier (0.0-1.0)"),
  created_at: z.string().default(() => new Date().toISOString()),
  is_verified: z.boolean().default(false),
});

export type Finding = z.infer<typeof FindingSchema>;

export function confidenceLabel(finding: Finding): string {
  if (finding.confidence >= 0.9) return "high";
  if (finding.confidence >= 0.75) return "medium";
  if (finding.confidence > 0) return "low";
  return "unverified";
}

export const ContradictionSchema = z.object({
  contradiction_id: z.string(),
  finding_id_a: z.string(),
  finding_id_b: z.string(),
  claim_a: z.string().default(""),
  claim_b: z.string().default(""),
  similarity: z
    .number()
    .min(0)
    .max(1)
    .default(0)
    .describe("Cosine similarity between the two claims (low = more contradictory)"),
  description: z.string().default(""),
  confirmed: z.boolean().default(false),
  created_at: z.string().default(() => new Date().toISOString()),
});

export type Contradiction = z.infer<typeof ContradictionSchema>;

export function severityLabel(contradiction: Contradiction): string {
  if (contradiction.similarity < 0.05) return "FATAL";
  if (contradiction.similarity < 0.1) return "MAJOR";
  return "MINOR";
}

export const GapSchema = z.object({
  gap_id: z.string(),
  description: z.string(),
  sub_question: z.string().optional(),
  severity: z.enum(["minor", "major", "critical"]).default("minor"),
  related_findings: z.array(z.string()).default([]),
});

export type Gap = z.infer<typeof GapSchema>;